templates/contribute/style.md

```diff
@@ -116,157 +98,171 @@ See our [documentation requirements](doc.html) for more suggestions and examples
-list.rec_on l
-  (assume H : x ∈ [], false.elim (iff.elim_left (mem_nil_iff _) H))
-  (assume y l,
-    assume IH : x ∈ l → ∃ s t : list T, l = s ++ (x::t),
-    assume H : x ∈ y::l,
-    or.elim (eq_or_mem_of_mem_cons H)
-      (assume H1 : x = y,
-        exists.intro [] (exists.intro l (by rw H1; refl)))
-      (assume H1 : x ∈ l,
-        let ⟨s, (H2 : ∃ t : list T, l = s ++ (x::t))⟩ := IH H1,
-            ⟨t, (H3 : l = s ++ (x::t))⟩ := H2 in
-        have H4 : y :: l = (y::s) ++ (x::t), by rw H3; refl,
-        exists.intro (y::s) (exists.intro t H4)))
+open List
+variable {T : Type}
 
+theorem mem_split {x : T} {l : List T} : x ∈ l → ∃ s t : List T, l = s ++ (x :: t) :=
+  List.recOn l
+    (fun H : x ∈ [] ↦ False.elim ((mem_nil_iff _).mp H))
+    (fun y l ↦
+      fun IH : x ∈ l → ∃ s t : List T, l = s ++ (x :: t) ↦
+      fun H : x ∈ y :: l ↦
+      Or.elim (eq_or_mem_of_mem_cons H)
+        (fun H1 : x = y ↦
+          Exists.intro [] (Exists.intro l (by rw [H1]; rfl)))
+        (fun H1 : x ∈ l ↦ 
+          let ⟨s, (H2 : ∃ t : List T, l = s ++ (x :: t))⟩ := IH H1
+          let ⟨t, (H3 : l = s ++ (x :: t))⟩ := H2
+          have H4 : y  ::  l = (y :: s) ++ (x :: t) := by rw [H3]; rfl
+          Exists.intro (y :: s) (Exists.intro t H4))) 
 ```
 
 A short declaration can be written on a single line:
 ```lean
-open nat
-lemma succ_pos : ∀ n : ℕ, 0 < succ n := zero_lt_succ
+open Nat
+theorem succ_pos : ∀ n : Nat, 0 < succ n := zero_lt_succ
 
-def square (x : nat) : nat := x * x
+def square (x : Nat) : Nat := x * x
 ```
 
-A "have" / "from" pair can be put on the same line.
+A `have` can be put on a single line when the justification is short.
 ```lean
-have H2 : n ≠ succ k, from subst (ne_symm (succ_ne_zero k)) (symm H),
-[...]
+example (n k : Nat) (h : n < k) : ... :=
+  have h1 : n ≠ k := ne_of_lt h
+  ...
 ```
-You can also put it on the next line, if the justification is long.
+When the justification is too long, you should put it on the next line, 
+indented by an additional two spaces.
 ```lean
-have H2 : n ≠ succ k,
-  from subst (ne_symm (succ_ne_zero k)) (symm H),
-[...]
+example (n k : Nat) (h : n < k) : ... :=
+  have h1 : n ≠ k :=
+    ne_of_lt h
+  ...
 ```
-If the justification takes more than a single line, keep the "from" on the same
-line as the "have", and then begin the justification indented on the next line.
+When the justification of the `have` uses tactic mode, the `by` should
+be placed on the same line, regardless of whether the justification
+spans multiple lines.
 ```lean
-have n ≠ succ k, from
-  not_intro
-    (assume H4 : n = succ k,
-      have H5 : succ l = succ k, from trans (symm H) H4,
-      have H6 : l = k, from succ_inj H5,
-      absurd H6 H2)))),
-[...]
+example (n k : Nat) (h : n < k) : ... :=
+  have h1 : n ≠ k := by exact ne_of_lt h
+  ...
+
+example (n k : Nat) (h : n < k) : ... :=
+  have h1 : n ≠ k := by
+    exact ne_of_lt h
+  ...
```

Comment:
It would be nice to have an example which doesn't encourage unnecessary `by exact`.